### PR TITLE
Break `docker-compose.yml` into smaller `compose.yml` files

### LINKED ISSUE #4111: Break down `docker-compose.yml` into smaller more-maintainable chunks
## Problem

In the current state, the `docker-compose.yml` file contains about 18 services and 7 volumes spanning the catalog, ingestion server, API and frontend. Put together the file is ~350 LoC which makes it quite cumbersome and coupled with #1910, the file is quite confusing.

## Description

Docker Compose supports breaking down the file in 3 ways.

- merging (using multiple files with `-f` flags)
- [extend](https://docs.docker.com/compose/multiple-compose-files/extends/)
- [include](https://docs.docker.com/compose/multiple-compose-files/include/)

By making good use of extend and include, we can _"compose"_ the Docker Compose config as a collection of smaller, more maintainable modules.

## Alternatives

We continue to use the same Docker Compose file but at least rorganise its contents to separate various layers of the stack into groups.

### PR DESCRIPTION
**Pull request by dhruvkb**
## Fixes

Fixes #4111 by @dhruvkb

## Description

This PR breaks down the main `docker-compose.yml` file into smaller `compose.yml` files placed in each of the stack folders. The shared services (like databases) and services not 100% related to any stack (like Plausible stuff) have been retained in the main file.

Additionally the ingestion server and indexer worker also follow the extension approach used by the Airflow services.

The ingestion server can no longer copy over the `docker-compose.yml` file and make edits to it, so it now uses `docker compose config` to let Docker Compose compile the configs (process `include`-d files, interpolate environment variables, normal all configs to the long form etc.) and uses that. This removes a few steps from the compose generation step.

## Testing Instructions

CI checks are passing so that should indicate that Docker Compose services are functional for the most part.

0. Run `just up` and `just ps` to see the state of the services.
1. Check out the repo.
2. Build all the images `just dc build`.
3. Run all the services `just dc up --force-recreate`.

## Checklist

- [x] My pull request has a descriptive title (not a vague title like`Update index.md`).
- [x] My pull request targets the _default_ branch of the repository (`main`) or a parent feature branch.
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [ ] I added or updated tests for the changes I made (if applicable).
- [x] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no visible errors.
- [ ] I ran the DAG documentation generator (if applicable).

[best_practices]:
  https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

**Comment by AetherUnbound** [comment]
> services not 100% related to any stack (like Plausible stuff)

Doesn't only the frontend use Plausible? What would prevent us from moving it to that division?

**Review by AetherUnbound — COMMENTED** [review]
This looks good! I'd recommend a few more changes:
- Plausible gets moved to frontend
- Minio/S3 gets moved to catalog

Otherwise I ran this locally well and everything worked as expected!

**Comment by dhruvkb** [comment]
There were a few analytics events that were going to be sent by the API. I don't know the status on those but we definitely had plans to use Plausible for API analytics as well at some point. Currently that is not the case, so I would be open to moving Plausible into `frontend/compose.yml` if that's the consensus.

File: catalog/compose.yml [hunk 1]
```diff
@@ -0,0 +1,67 @@
+# Common build configuration for Airflow
```
[the remark this answers is not in the document]
Comment (dhruvkb) [reply catalog/compose.yml:1]: Can you also confirm the env vars in the root env template can be moved to `catalog/env.template`? Only the catalog services use them.

**Review by sarayourfriend — APPROVED** [review]
LGTM! I don't think we necessarily need to move anything that's "general" infrastructure (Redis, DBs, Plausible), but it's easy enough to move things one place or another later if it makes sense (or stops make sense), so it isn't a blocker to me either way.

**Review by krysal — APPROVED** [review]
A welcome change! This file was certainly getting very large. LGTM!

**Pull request merged**